crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs

```diff
@@ -13,15 +13,67 @@
 // limitations under the License.
 
 use indexed_db_futures::{prelude::*, web_sys::DomException};
-use tracing::info;
+use matrix_sdk_crypto::olm::InboundGroupSession;
+use tracing::{debug, info};
 use wasm_bindgen::JsValue;
 
-use crate::crypto_store::{keys, Result};
+use crate::{
+    crypto_store::{
+        indexeddb_serializer::IndexeddbSerializer, keys, InboundGroupSessionIndexedDbObject, Result,
+    },
+    IndexeddbCryptoStoreError,
+};
+
+mod old_keys {
+    /// Old format of the inbound_group_sessions store which lacked indexes or a
+    /// sensible structure
+    pub const INBOUND_GROUP_SESSIONS_OLD: &str = "inbound_group_sessions";
+}
 
 /// Open the indexeddb with the given name, upgrading it to the latest version
 /// of the schema if necessary.
-pub async fn open_and_upgrade_db(name: &str) -> Result<IdbDatabase, DomException> {
-    let mut db_req: OpenDbRequest = IdbDatabase::open_u32(name, 5)?;
+pub async fn open_and_upgrade_db(
+    name: &str,
+    serializer: &IndexeddbSerializer,
+) -> Result<IdbDatabase, IndexeddbCryptoStoreError> {
+    // This is all a bit of a hack. Some of the version migrations require a data
+    // migration, which has to be done via async APIs; however, the
+    // JS `upgrade_needed` mechanism does not allow for async calls.
+    //
+    // Start by finding out what the existing version is, if any.
+    let db = IdbDatabase::open(name)?.await?;
+    let old_version = db.version() as u32;
+    db.close();
+
+    // If we have yet to complete the migration to V7, migrate the schema to V6
+    // (if necessary), and then migrate any remaining data.
+    if old_version <= 6 {
+        let db = migrate_schema_to_v6(name).await?;
+        migrate_data_for_v6(serializer, &db).await?;
+        db.close();
+    }
+
+    // Now we can safely complete the migration to V7 which will drop the old store.
+    let mut db_req: OpenDbRequest = IdbDatabase::open_u32(name, 7)?;
+    db_req.set_on_upgrade_needed(Some(|evt: &IdbVersionChangeEvent| -> Result<(), JsValue> {
+        let old_version = evt.old_version() as u32;
+        let new_version = evt.new_version() as u32;
+
+        info!(old_version, new_version, "Upgrading IndexeddbCryptoStore, phase 2");
```

Comment:
I don't like the idea of spreading "phase" numbers across different functions, especially since it's not obvious in this function what phase 1 is, plus we'd have to maintain forward-looking consistency e.g. if we ever introduced a phase 3, reference it consistently.

A lot of words to say, maybe we can tell "Getting back to schema migrations" here?